doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -1,209 +1,213 @@
-   replaces every occurrence of `term`:subscript:`1`' by `term`:subscript:`2`'.
-   Hence, some of the variables :g:`x`\ :sub:`i` are solved by unification,
-   and some of the types :g:`A`\ :sub:`1`:g:`, ..., A`\ :sub:`n` become new
-   subgoals.
+  .. todo PR: A and B are convertible iff they are definitionally equal??
 
-   .. exn:: The @term provided does not end with an equation.
-      :undocumented:
+.. _rewritingexpressions:
 
-   .. exn:: Tactic generated a subgoal identical to the original goal. This happens if @term does not occur in the goal.
-      :undocumented:
+Rewriting with Leibniz and setoid equality
+------------------------------------------
 
-   .. tacv:: rewrite -> @term
+.. tacn:: rewrite {+, @oriented_rewriter } {? @occurrences } {? by @ltac_expr3 }
 
-      Is equivalent to :n:`rewrite @term`
+   .. insertprodn oriented_rewriter one_term_with_bindings
 
-   .. tacv:: rewrite <- @term
+   .. prodn::
+      oriented_rewriter ::= {? {| -> | <- } } {? @natural } {? {| ? | ! } } @one_term_with_bindings
+      one_term_with_bindings ::= {? > } @one_term {? with @bindings }
 
-      Uses the equality :n:`@term`:sub:`1` :n:`= @term` :sub:`2` from right to left
+   .. todo PR: When I search for "rewrite (" in the .v files, I see tons of examples that I don't
+      understand.  Need to say something about them.
 
-   .. tacv:: rewrite @term in @goal_occurrences
+   Rewrites terms based on equalities.  :n:`@one_term` can be the name of a hypothesis that has
+   the form :n:`EQ @term__1 @term__2` or a lemma that has the form:
 
-      Analogous to :n:`rewrite @term` but rewriting is done following
-      the clause :token:`goal_occurrences`. For instance:
+   :n:`forall (x__1: A__1) ... (x__n: A__n), EQ @term__1 @term__2`
 
-      + :n:`rewrite H in H'` will rewrite `H` in the hypothesis
-        ``H'`` instead of the current goal.
-      + :n:`rewrite H in H' at 1, H'' at - 2 |- *` means
-        :n:`rewrite H; rewrite H in H' at 1; rewrite H in H'' at - 2.`
-        In particular a failure will happen if any of these three simpler tactics
-        fails.
-      + :n:`rewrite H in * |-` will do :n:`rewrite H in H'` for all hypotheses
-        :g:`H'` different from :g:`H`.
-        A success will happen as soon as at least one of these simpler tactics succeeds.
-      + :n:`rewrite H in *` is a combination of :n:`rewrite H` and :n:`rewrite H in * |-`
-        that succeeds if at least one of these two tactics succeeds.
+   where :g:`EQ` is the Leibniz equality `eq` or a registered setoid equality.  (Typically
+   the infix notation :n:`@term__1 = @term__2` is defined as :n:`eq @term__1 @term__2`.
+   Either form may used.)
 
-      Orientation :g:`->` or :g:`<-` can be inserted before the :token:`term` to rewrite.
+   .. todo PR: I can't remember where the following sentence came from.  It seems
+      incorrect.  Maybe worth having some comparison to setoid_rewrite, though
```

Comment:
Indeed, it is incorrect. `rewrite` can use the default implementation for Leibniz equality or the setoid rewrite implementation automatically. However, the user may also choose to call the latter explicitly even in the case of a Leibniz equality by using the `setoid_rewrite` tactic. If `rewrite` picks the setoid implementation, it will fail with an error message if the `Setoid` library hasn't been required first though.

A good way of finding out what implementation is used is to set the `Typeclasses Debug` flag (since setoid rewrite relies on the typeclass mechanism internally):

```coq
Require Setoid.

Set Typeclasses Debug.
Axiom foo : True <-> False.
Goal True.
rewrite foo.
(* Use the setoid implementation since the output is:
1: looking for (RelationClasses.subrelation iff (Basics.flip Basics.impl)) without backtracking
1.1: exact Morphisms.iff_flip_impl_subrelation on (RelationClasses.subrelation iff (Basics.flip Basics.impl)), 0 subgoal(s)
Finished run 1 of resolution.
*)
Abort.


Axiom bar : 0 = 1.
Goal 0 = 0.
rewrite bar.
(* Use the default implementation since there is no output. *)
Abort.